### PR TITLE
For nsdi opportunistic

### PR DESCRIPTION
**Pull request by KenBirman**
Looking at the new code to see if I can help localize the bug Lorenzo is struggling with.  I'm puzzled about why GitHub is claiming that I want to merge these into master -- I wasn't even going to change the nsdi-opportunistic code, just to review it.

### CODE REVIEW

**Comment by songweijia** [comment]
I remember Ken and Lorenzo is working on an issue SST is called before it is initialized. Should we merge after that issue is solved and tested with our regression test framework?

**Comment by KenBirman** [comment]
I completely agree!  In fact I do want Lorenzo's changes merged in... but not until that bug is fixed.  I don't fully understand what I did that caused GitHub to create this "merge desired" thread (but because we need Lorenzo's optimizations, I didn't just delete the thread immediately).

My worry is that Lorenzo's bug is a rather serious core Derecho race condition that probably exists in the original system but simply wasn't noticed.  Lorenzo and Sagar did add these additional "index" fields that need to be initialized to -1, and this revealed that something is pushing SST rows before initialization is finished.  But Edward also has some fields that must be initialized in the SST.  So it is important to understand what the cause here might be.

It may be as simple as when the initializer runs (it should run before Edward's SST setup, but if somehow Edward's SST logic were to run first, then when it finishes and pushes the initialized rows, bam, we would have our bug).

**Comment by KenBirman** [comment]
The Travis tests crash when connecting to a spdlog server on some kind of mirror website.  Does anyone know what role this mirror site plays for spdlog?  I thought we were using it purely locally...

**Comment by KenBirman** [comment]
The Travis build crashes when connecting to spdlog on a mirror website.  Does anyone know what role this plays?  I thought we used spdlog locally.

The failed action is an attempted connection to:

http://mirrors.edge.kernel.org/ubuntu/pool/universe/s/spdlog/libspdlog-dev_1.3.1-1_amd64.deb

**Comment by sagarjha** [comment]
The spdlog library is installed for local testing on the TravisCI machine through that link. It seems that on the website, http://mirrors.kernel.org/ubuntu/pool/universe/s/spdlog/, they updated the spdlog versions from 1.3 to 1.5 and 1.6. Therefore, the deb file we are seeking is not there anymore.

**Comment by ellerre** [comment]
Just a short description of what this branch contains and why we merge it. This branch contains important performance improvements to the SMC protocol. The atomic multicast bandwidth utilization for small messages (up to 10KB) improves up to 10x compared to a baseline implementation. The main improvements are:
-  Memory polling optimizations using an opportunistic batching techniques at all the steps of message processing: send, receive and delivery. 
- Improvements to null-message send logic: in case of one or more null messages in a batch, we only send the header of the first one.
- Improvements to multi-thread synchronization: locks are not held when issuing RDMA operations.

File: src/core/view_manager.cpp [hunk 1]
```diff
@@ -923,10 +926,10 @@ void ViewManager::propose_changes(DerechoSST& gmsSST) {
         gmssst::set(gmsSST.changes[my_rank][last_change_index].end_of_view, false);
     } else {
         //Push all the proposed changes, including joiner information if any joins were proposed
-        gmsSST.put(gmsSST.changes);
+        gmsSST.put_with_completion(gmsSST.changes);
```
Comment (sagarjha) [inline src/core/view_manager.cpp:929]: I don't see why we need to change the put call to put_with_completion. Are you just trying to safe? These protocols work by the nodes acknowledging these changes in multiple steps. So waiting for completion is redundant.

File: src/core/view_manager.cpp [hunk 2]
```diff
@@ -923,10 +926,10 @@ void ViewManager::propose_changes(DerechoSST& gmsSST) {
...
         if(!proposed_join_sockets.empty()) {
-            gmsSST.put(gmsSST.joiner_ips.get_base() - gmsSST.getBaseAddress(),
-                       gmsSST.num_changes.get_base() - gmsSST.joiner_ips.get_base());
+            gmsSST.put_with_completion(gmsSST.joiner_ips.get_base() - gmsSST.getBaseAddress(),
```
Comment (sagarjha) [inline src/core/view_manager.cpp:931]: same here

File: src/core/view_manager.cpp [hunk 3]
```diff
@@ -1052,8 +1055,8 @@ void ViewManager::new_leader_takeover(DerechoSST& gmsSST) {
         make_subgroup_maps(subgroup_info, curr_view, *next_view);
         //Push the entire new changes vector and the associated joiner_ip vectors
         gmsSST.put(gmsSST.changes);
-        gmsSST.put(gmsSST.joiner_ips.get_base() - gmsSST.getBaseAddress(),
-                   gmsSST.num_changes.get_base() - gmsSST.joiner_ips.get_base());
+        gmsSST.put_with_completion(gmsSST.joiner_ips.get_base() - gmsSST.getBaseAddress(),
```
Comment (sagarjha) [inline src/core/view_manager.cpp:1058]: and here

File: src/core/view_manager.cpp [hunk 4]
```diff
@@ -1106,7 +1109,7 @@ void ViewManager::acknowledge_proposed_change(DerechoSST& gmsSST) {
         //This pushes the contiguous set of joiner_xxx_ports fields all at once
         gmsSST.put(gmsSST.joiner_ips.get_base() - gmsSST.getBaseAddress(),
                    gmsSST.num_changes.get_base() - gmsSST.joiner_ips.get_base());
-        gmsSST.put(gmsSST.num_changes);
+        gmsSST.put_with_completion(gmsSST.num_changes);
```
Comment (sagarjha) [inline src/core/view_manager.cpp:1112]: here as well

**Review by sagarjha — CHANGES_REQUESTED** [review]
Lorenzo, the merge looks good to me. I have just one minor issue that regards changing put calls to put_with_completion calls. Let me know why you changed it that way.

**Comment by ellerre** [comment]
Hi Sagar, 
those are changes introduced by @etremel in commit 866a0e04d9266f10e0cbe6f6ae2355cc77317f29 , when we were trying to understand the cause of the initialization bug. Now that we fixed it, I guess we could go back to the previous version.

**Comment by sagarjha** [comment]
I see. I am adding a commit to revert those calls back to simple put.

**Comment by sagarjha** [comment]
In function terminate_epoch, we still have calls to put_with_completion of the entire row. Should we also not push the slots in that function?

**Comment by KenBirman** [comment]
Hold on -- Edward also had other issues that could relate to this.  He was using the asynchronous put (without completion) all through his logic, but in fact on IB that would be incorrect; his SST vector fields need to be sent in a sequence in which the last put is done with a completion, before he pushes the updated guard (counter).  Otherwise it works on LF but when mapped directly to verbs, he would have no memory fence between updating the guard and updating the vector

**Comment by KenBirman** [comment]
> I see. I am adding a commit to revert those calls back to simple put.

I'm not sure -- as noted, Edward was improperly omitting a fence between pushing his vector of proposed updates and pushing the counter for that list of updates.  The put-with-completions provides this necessary barrier.

**Comment by sagarjha** [comment]
okay, I am holding on. Are we still assuming that two verbs, one after the other, are separated by a memory fence?

**Comment by KenBirman** [comment]
If verbs A and then B are posted, and A is posted with completion, there is a memory fence between A and B.  But if A is posted without a completion, followed by B, they can be applied to memory concurrently by the DMA controller of the target NIC.  So, in that case, no fence.

**Comment by sagarjha** [comment]
> If verbs A and then B are posted, and A is posted with completion, there is a memory fence between A and B. But if A is posted without a completion, followed by B, they can be applied to memory concurrently by the DMA controller of the target NIC. So, in that case, no fence.

I am not sure about this. Completion only means that the sender node receives an entry in its completion queue. A completion is generated regardless and it has nothing to do with a barrier on the remote side. See this discussion at https://github.com/ofiwg/libfabric/pull/4475. @songweijia What do you think?

**Comment by KenBirman** [comment]
Using put with completion, the sender of A actually waits to see that the operation has completed before posting B.  This is a strong fence.

**Comment by sagarjha** [comment]
Of course, that is true. I am saying that we don't need it in this case. Even with verbs without completions, we have always assumed that they are applied sequentially on the remote side.
From the Derecho paper,
"
Accordingly, when updating multiple entries in a vector, we take advantage
of a different property: RDMA writes respect the sender’s FIFO ordering, in that multiple verbs
are applied at the target node sequentially. Thus, we can guard the vector within the SST with a
counter, provided that we update the vector first and then the counter in a separate verb. When
a reader sees the guard change, it is safe for it to read the guarded vector elements. It can then
acknowledge the data, if needed, via an update to its own SST row.
"

**Comment by KenBirman** [comment]
Ah, but since then Simon Peter told me that this is not a correct assumption.  It only holds if a completion is requested on the "earlier" of the operations.  Otherwise the modern semantic is that A and B can be performed concurrently unless they access the identical cache line (if A and B update the same cache line, you get sequential consistency, respecting the order in which they were posted).

**Comment by KenBirman** [comment]
For LibFabric this is all irrelevant since they always set the completion bit, hence the remote DMA controller will actually fence A before applying B.  But when we run directly on verbs, it matters.

**Comment by sagarjha** [comment]
We don't support directly running on verbs anymore. SMC actively makes use of the property that is true for libfabrics, and maybe not for verbs. If we were to change every SMC message send to wait for completion (otherwise, we cannot update the index of messages sent reliably), I am not sure about the resulting impact on performance.

**Comment by KenBirman** [comment]
Well, if we do not support verbs, that helps.  In that case because LibFabrics always requests a completion even when the one-sided write is issued asynchronously from code that is not not watching for that completion, we would be safe and correct without this change.

**Comment by etremel** [comment]
My reply might be a little moot at this point, but just to confirm what Ken and Lorenzo explained earlier: I changed those `put()`s  to `put_with_completion()` in response to Ken's observation that a sequence of one-sided writes needs at least one completion request before the final "guard" write in order to ensure that the final write actually occurs last on the receiver side. In each case where I made a change in view_manager.cpp, the protocol was only correct if the final write (for example, the `put(gmsSST.num_changes)`) appeared in the receiver's memory strictly after all of the other writes in the sequence (e.g. the new proposed change information).

If, in fact, we don't need to explicitly request completions when running on LibFabrics, and LibFabrics is our only supported platform now, then my changes can all be reverted back to simple `put()`s.

**Comment by sagarjha** [comment]
Thanks Edward. I have one final question to both @ellerre and @etremel - In function `finish_view_changes`, we have replaced a call to `put_with_completion` with `push_row_except_slots`. In the function `terminate_epoch`, we still push the entire row to the members twice with completion. I guess that is important because we want to deliver any left over SMC messages before moving on to the next view?

**Comment by ellerre** [comment]
Yes, that seems to be the reason. For each subgroup, we check whether there are pending SMC messages and then we push the entire row once to ensure all the messages are sent. After that, we force the reception of those messages. The last push, I guess, ensures that all the control plane messages of the last subgroup have been sent to the others. Probably @etremel can explain this much better than me.

**Comment by KenBirman** [comment]
I'm thinking about this question and it strikes me that we may have the same issue here that Lorenzo ran into during initialization!

If the SST row is huge (which can easily happen if SMC uses fairly large message limits, has many slots, and on top of that we have many subgroups -- like in the DDS case), the SMC portion of the row could be tens or hundreds of megabytes.   

Yet because these slots are always pushed by the row owner (and only by the row owner) when an SMC multicast is sent, there is NEVER a reason to include them in an sst put.  Basically, we should replace all SST full-row puts with a sequence, like Lorenzo did: the portion "below" the SMC slots, and then the portion "above" the SMC slots.  Instead of megabytes, our biggest transfer would be maybe 1K.

Could I suggest that we actually go ahead and make this change?  It seems like it would have a really dramatic impact on performance of the epoch termination code, especially for the DDS (which is our main user!)

**Comment by KenBirman** [comment]
One more remark: If the end of epoch logic also uses the TCP-based SST sync barrier, we need to be sure that the last put done prior to calling that SST sync method definitely requested a completion, and waited for it.

**Comment by ellerre** [comment]
In this branch, we already removed all the calls to _put()_ without arguments. We have only two calls to _put_with_completion()_ in the method Sagar mentioned, i.e., ViewManager::terminate_epoch(). If I understand correctly, this method ensures that all the SMC messages sent by the user application (_pending_ messages) are actually issued on the wire and acknowledged before the epoch termination. Hence in this case it seems that we do need to send the SST portion with the slots.

**Comment by KenBirman** [comment]
Ok, that makes sense.

**Review by sagarjha — APPROVED** [review]
I am finished with my review. This is ready to be submitted. Lorenzo, I know you were testing performance of this new version. If you have any changes from those tests, you can make them. Otherwise, go ahead and finish the merge.

I have create a new issue to track pushing the SMC slots - https://github.com/Derecho-Project/derecho/issues/188. We can handle that later in the future.

**Comment by ellerre** [comment]
Thank you, Sagar. Thanks also to the whole Derecho team for the help with this pull request.

**Review by ellerre — APPROVED** [review]

**Pull request merged**